Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -732,36 +733,44 @@ public struct ObjectId: BSONValue, Equatable, CustomStringConvertible, Codable {
     public var bsonType: BSONType { return .objectId }
 
     /// This `ObjectId`'s data represented as a `String`.
-    public let oid: String
+    public var hex: String {
+        var str = Data(count: 25)
+        return str.withUnsafeMutableBytes { (rawBuffer: UnsafeMutablePointer<Int8>) in
+            withUnsafePointer(to: self.oid) { oidPtr in
+                bson_oid_to_string(oidPtr, rawBuffer)
+            }
+            return String(cString: rawBuffer)
+        }
+    }
 
     /// The timestamp used to create this `ObjectId`
-    public let timestamp: UInt32
+    public var timestamp: UInt32 {
+        return withUnsafePointer(to: self.oid) { oidPtr in UInt32(bson_oid_get_time_t(oidPtr)) }
+    }
 
-    /// Initializes a new `ObjectId`.
-    public init() {
-        var oid_t = bson_oid_t()
-        bson_oid_init(&oid_t, nil)
-        self.init(fromPointer: &oid_t)
+    public var description: String {
+        return self.hex
     }
 
-    /// Initializes an `ObjectId` from the provided `String`. Assumes that the given string is a valid ObjectId.
-    /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
-    public init(fromString oid: String) {
+    internal let oid: bson_oid_t
+
+    /// Initializes a new `ObjectId`.
+    public init() {
+        var oid = bson_oid_t()
+        bson_oid_init(&oid, nil)
         self.oid = oid
-        var oid_t = bson_oid_t()
-        bson_oid_init_from_string(&oid_t, oid)
-        self.timestamp = UInt32(bson_oid_get_time_t(&oid_t))
     }
 
     /// Initializes an `ObjectId` from the provided `String`. Returns `nil` if the string is not a valid
```

Comment:
maybe we should say it's a hex string here?

Reply:
done